### PR TITLE
feature(search functionality): Implementing search functionality and custom filtering

### PR DESCRIPTION
**Pull request by Etomovich**
**Description**
<hr/>

This PR adds custom filtering and search functionality. A user can be able to perform partial and full word searches depending on the `search string`  he/she wants to search. This feature also enables the user to filter his/her search depending on the specific fields he/she wants to search. Currently, the fields that the user can search from are `author`, `tags`, `title`, `description` and `body`.  The user will receive search results of all the filters in different tabs each for each filter allowing him/her to navigate through the all the filters of the same search.

**Type of Change**
<hr/>

- [x] feature( a breaking change that implementing search functionality and custom filtering.)

**Checklist**
<hr/>

- [x] Users should be able to search for articles.
- [x] Users should be able to filter articles according to the different fields
- [x] Add test to show search functionality operations.

**Related Stories**
<hr/>

[Implement search and custom filtering functionality](https://www.pivotaltracker.com/n/projects/2313280/stories/164046225)

### CODE REVIEW

File: src/store/actions/searchActions.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    axios.get(REACT_APP_BASE_URL.concat(`articles/title/search/?query=${searchString}`)),
+    axios.get(REACT_APP_BASE_URL.concat(`articles/author/search/?query=${searchString}`)),
+    axios.get(REACT_APP_BASE_URL.concat(`articles/description/search/?query=${searchString}`)),
+    axios.get(REACT_APP_BASE_URL.concat(`articles/tags/search/?query=${searchString}`)),
+  ], {
+    headers: {
+      'content-type': 'application/json',
+    },
+  },
+)
```
[the remark this answers is not in the document]
Comment (Etomovich) [reply src/store/actions/searchActions.js]: > Unexpected newline before ')' function-paren-newline

Fixed.

File: src/store/actions/searchActions.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { toast } from 'react-toastify';
+import axios from 'axios';
+import { GET_SEARCH_RESULTS } from './actionTypes';
+
+const { REACT_APP_BASE_URL } = process.env;
+
+export const fetchSearchResults = searchString => dispatch => axios.all(
```
[the remark this answers is not in the document]
Comment (Etomovich) [reply src/store/actions/searchActions.js]: > Unexpected newline after '(' function-paren-newline

Fixed.

File: src/containers/SearchArticle.jsx [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+                  loadMore={this.loadMoreArticles}
+                  hasMore={descriptionHasNext}
+                  loader={loader}
+                >
+                  <div className="tracks">
+                    {
+                    descriptionSearchResults.map(item => (
+                      <div key={item} className="search_results">
+                        <h3>
+                          <Link
```
[the remark this answers is not in the document]
Comment (Etomovich) [reply src/containers/SearchArticle.jsx]: Fixed.

[the remark this answers is not in the document]
Comment (Etomovich) [reply src/containers/SearchArticle.jsx:239]: > The href attribute is required for an anchor to be keyboard accessible. Provide a valid, navigable address as the href value. If you cannot provide an href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md jsx-a11y/anchor-is-valid

Fixed.

File: src/containers/SearchArticle.jsx [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+                  loadMore={this.loadMoreArticles}
+                  hasMore={bodyHasNext}
+                  loader={loader}
+                >
+                  <div className="tracks">
+                    {
+                    bodySearchResults.map(item => (
+                      <div key={item} className="search_results">
+                        <h3>
+                          <Link
```
[the remark this answers is not in the document]
Comment (Etomovich) [reply src/containers/SearchArticle.jsx:209]: > The href attribute is required for an anchor to be keyboard accessible. Provide a valid, navigable address as the href value. If you cannot provide an href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md jsx-a11y/anchor-is-valid

Fixed.

File: src/containers/SearchArticle.jsx [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+                  loadMore={this.loadMoreArticles}
+                  hasMore={tagsHasNext}
+                  loader={loader}
+                >
+                  <div className="tracks">
+                    {
+                    tagsSearchResults.map(item => (
+                      <div key={item} className="search_results">
+                        <h3>
+                          <Link
```
[the remark this answers is not in the document]
Comment (Etomovich) [reply src/containers/SearchArticle.jsx:179]: Fixed.

File: src/components/Header.jsx [hunk 6]
```diff
@@ -32,46 +34,47 @@ export class Header extends React.Component {
...
+                overlay={(
+                  <Popover className="user-profile-container ml-auto">
+                    <Link to="/article/create" className="nav-dropdown">New Article</Link>
+                    <br />
+                    <br />
+                    <Link to="/articles/drafts" className="nav-dropdown">Drafts</Link>
+                    <hr />
+                    <Link to="/profile" className="nav-dropdown">Profile</Link>
+                    <hr />
+                    <span className="logout nav-dropdown" onClick={this.dispatchLogout}>Logout</span>
```
[the remark this answers is not in the document]
Comment (Etomovich) [reply src/components/Header.jsx]: > Static HTML elements with event handlers require a role jsx-a11y/no-static-element-interactions
> Visible, non-interactive elements with click handlers must have at least one keyboard listener jsx-a11y/click-events-have-key-events

Fixed.

File: src/components/Header.jsx [hunk 7]
```diff
@@ -32,46 +34,47 @@ export class Header extends React.Component {
...
+                trigger="click"
+                placement="bottom"
+                overlay={(
+                  <Popover className="user-profile-container ml-auto">
+                    <Link to="/article/create" className="nav-dropdown">New Article</Link>
+                    <br />
+                    <br />
+                    <Link to="/articles/drafts" className="nav-dropdown">Drafts</Link>
+                    <hr />
+                    <Link to="/profile" className="nav-dropdown">Profile</Link>
```
[the remark this answers is not in the document]
Comment (Etomovich) [reply src/components/Header.jsx]: > The href attribute is required for an anchor to be keyboard accessible. Provide a valid, navigable address as the href value. If you cannot provide an href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md jsx-a11y/anchor-is-valid

Fixed.

File: src/components/Header.jsx [hunk 8]
```diff
@@ -32,46 +34,47 @@ export class Header extends React.Component {
...
+            {authenticated ? (
+              <OverlayTrigger
+                trigger="click"
+                placement="bottom"
+                overlay={(
+                  <Popover className="user-profile-container ml-auto">
+                    <Link to="/article/create" className="nav-dropdown">New Article</Link>
+                    <br />
+                    <br />
+                    <Link to="/articles/drafts" className="nav-dropdown">Drafts</Link>
```
[the remark this answers is not in the document]
Comment (Etomovich) [reply src/components/Header.jsx]: Fixed.

File: src/components/Header.jsx [hunk 9]
```diff
@@ -32,46 +34,47 @@ export class Header extends React.Component {
...
+            <Nav.Item className="search-placement">
+              <SearchForm />
+            </Nav.Item>
+            {authenticated ? (
+              <OverlayTrigger
+                trigger="click"
+                placement="bottom"
+                overlay={(
+                  <Popover className="user-profile-container ml-auto">
+                    <Link to="/article/create" className="nav-dropdown">New Article</Link>
```
[the remark this answers is not in the document]
Comment (Etomovich) [reply src/components/Header.jsx]: Fixed.

> The href attribute is required for an anchor to be keyboard accessible. Provide a valid, navigable address as the href value. If you cannot provide an href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md jsx-a11y/anchor-is-valid

Fixed.

File: src/components/Header.jsx [hunk 10]
```diff
@@ -32,46 +34,47 @@ export class Header extends React.Component {
...
+                    <hr />
+                    <span className="logout nav-dropdown" onClick={this.dispatchLogout}>Logout</span>
+                  </Popover>
+              )}
+                rootClose
               >
+                <img className="rounded-circle float-right space-icon" src="https://marriedbiography.com/wp-content/uploads/2018/01/Enrique-Iglesias.jpg" alt="user" />
+              </OverlayTrigger>
+            ) : (
+              <>
```
[the remark this answers is not in the document]
Comment (abulojoshua1) [reply src/components/Header.jsx]: Why do you have empty tags here `<>`

File: src/components/Header.jsx [hunk 11]
```diff
@@ -32,46 +34,71 @@ export class Header extends React.Component {
...
+                  <Popover className="user-profile-container ml-auto">
+                    <Link
+                      to="/article/create"
+                      href="/article/create"
+                      className="nav-dropdown"
+                    >
+                      New Article
+                    </Link>
+                    <br />
+                    <br />
```
Comment (abulojoshua1) [inline src/components/Header.jsx]: why are you using <br /> twice here?

File: src/components/Header.jsx [hunk 12]
```diff
@@ -32,46 +34,71 @@ export class Header extends React.Component {
...
+                >
+                  Login
+                </Nav.Item>
+                <Nav.Item
+                  onClick={this.dispatchRegister}
+                  className="nav-link auth-btn get-started"
+                >
+                  Get Started
+                </Nav.Item>
+              </>
```
Comment (abulojoshua1) [inline src/components/Header.jsx]: why do you have empty tags here `</>`?

**Review by abulojoshua1 — CHANGES_REQUESTED** [review]

File: src/components/Header.jsx (hunk 11 above)
Comment (Etomovich) [reply src/components/Header.jsx]: > why are you using 
> twice here?

This has been handled.

**Review by abulojoshua1 — APPROVED** [review]
LGTM

File: src/components/Header.jsx (hunk 12 above)
Comment (Etomovich) [reply src/components/Header.jsx]: Replaced this with full `<React.Fragment>`. I think hound CI does not recognize the shorthand method.

**Review by abulojoshua1 — APPROVED** [review]
LGTM

**Pull request merged**